hazelcast/src/test/java/com/hazelcast/nio/DeserializationCheckerTest.java

```diff
@@ -0,0 +1,143 @@
+    public void testPackageDefaultBlacklisted() throws ClassNotFoundException {
+        DeserializationChecker.checkClassNameForResolution("org.apache.commons.collections.functors.Test");
+    }
+
+    /**
+     * <pre>
+     * Given: Blacklist is set to empty string (i.e. blacklisting disabled).
+     * When: {@link DeserializationChecker#checkClassNameForResolution(String)} is called for a class included in the default blacklist.
+     * Then: no exception is thrown
+     * </pre>
+     */
+    @Test
+    public void testEmptyBlacklist() throws ClassNotFoundException {
+        ruleSysPropBlacklist.setOrClear("");
+        DeserializationChecker.checkClassNameForResolution("bsh.XThis");
+    }
+
+    /**
+     * <pre>
+     * Given: Whitelist is set.
+     * When: {@link DeserializationChecker#checkClassNameForResolution(String)} is called for a whitelisted class.
+     * Then: no exception is thrown
+     * </pre>
+     */
+    @Test
+    public void testClassInWhitelist() throws ClassNotFoundException {
+        ruleSysPropWhitelist.setOrClear("java.lang.Test1,java.lang.Test2,java.lang.Test3");
+        DeserializationChecker.checkClassNameForResolution("java.lang.Test2");
+    }
+
+    /**
+     * <pre>
+     * Given: Whitelist is set.
+     * When: {@link DeserializationChecker#checkClassNameForResolution(String)} is called for a class which has whitelisted package.
+     * Then: no exception is thrown
+     * </pre>
+     */
+    @Test
+    public void testPackageInWhitelist() throws ClassNotFoundException {
+        ruleSysPropWhitelist.setOrClear("java.lang.Test1,com.whitelisted");
+        DeserializationChecker.checkClassNameForResolution("com.whitelisted.Test2");
+    }
+
+    /**
+     * <pre>
+     * Given: Whitelist is set.
+     * When: {@link DeserializationChecker#checkClassNameForResolution(String)} is called for a not whitelisted class.
+     * Then: {@link ClassNotFoundException} is thrown
+     * </pre>
+     */
+    @Test(expected=ClassNotFoundException.class)
+    public void testClassNotInWhitelist() throws ClassNotFoundException {
+        ruleSysPropWhitelist.setOrClear("java.lang.Test1,java.lang.Test2,java.lang.Test3");
+        DeserializationChecker.checkClassNameForResolution("java.lang.Test4");
+    }
+
+    /**
+     * <pre>
+     * Given: Blacklist and Whitelist are set.
+     * When: {@link DeserializationChecker#checkClassNameForResolution(String)} is called for a class which is whitelisted and blacklisted together.
+     * Then: {@link ClassNotFoundException} is thrown
+     * </pre>
+     */
+    @Test(expected=ClassNotFoundException.class)
+    public void testWhitelistedAndBlacklisted() throws ClassNotFoundException {
+        ruleSysPropWhitelist.setOrClear("java.lang.Test1,java.lang.Test2,java.lang.Test3");
+        ruleSysPropBlacklist.setOrClear("java.lang.Test3,java.lang.Test2,java.lang.Test1");
+        DeserializationChecker.checkClassNameForResolution("java.lang.Test1");
+    }
+}
```

Comment:
Missing newline at EOF

Reply:
Hi guys, original reporter here! If a whitelist is present, only classes on the union of all whitelists should be allowed. If no whitelist is present, classes not present on the union of all blacklists should be allowed. In both cases, any defaults should be merged with any additions by user configuration. The user should be required to explicitly remove items on the default blacklist. 

For example, Weblogic allows a `-class` syntax in user filters which removes any matching classes or packages from the list.

I also agree that `ClassNotFoundException` isn't the right exception to throw here, because it doesn't communicate the correct information. The exception message back to the client should include something indicating that deserialization failed due to filtering, but not give too much information. If local logging is configured, the local log should include more information.

Reply:
Or more accurately:

If (a class matches a pattern in the whitelist or the whitelist is empty), and it does not also match a pattern in the blacklist, then it is allowed.

If a class matches a pattern in the blacklist, and it does not also match a pattern in the whitelist, then it is not allowed.

Else, it is allowed.

This allows whitelisting `com.package.*`, then disallowing specific classes in the package.